### PR TITLE
Add ROS Interface to handle schedules

### PR DESCRIPTION
**Pull request by sktometometo**
Add ROS Interface to handle schedules.

- Add ROS message and services
- Add ROS Interface
  + Topics
    - `~app_schedules` : list of schedules
  + Services
    - `~add_entry` : Add a schedule
    - `~remove_entry` : Remove schedule

### CODE REVIEW

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 1]
```diff
@@ -24,15 +31,61 @@ def __init__(self, robot_name, yaml_path, duration, update_duration):
...
+            app['app_schedule']['start'] = entry.app_schedule.start
+        if entry.app_schedule.stop != '':
+            app['app_schedule']['stop'] = entry.app_schedule.stop
+        rospy.loginfo(
+            'register app schedule => name: {0}, app_name: {1}'.format(
+                app['name'], app['app_name']))
+        self.app_lock.acquire()
+        self.apps.append(app)
+        self._register_app(app)
+        self.app_lock.release()
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]: I think it is better to use `with` for `lock`.
https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 2]
```diff
@@ -55,20 +108,23 @@ def _register_app(self, app):
             app_args = []
         start_job = self._create_start_job(name, app_name, app_args)  # NOQA
         try:
-            eval('schedule.{}.do(start_job)'.format(app_schedule['start']))
+            eval('schedule.{}.do(start_job).tag(\'{}\')'.format(app_schedule['start'],app['name']))
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
            eval('schedule.{}.do(start_job).tag(\'{}\')'.format(app_schedule['start'], app['name']))
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 3]
```diff
@@ -55,20 +108,23 @@ def _register_app(self, app):
...
         except (AssertionError, ValueError) as e:
             rospy.logerr(e)
             rospy.logerr('Cannot register start app')
             rospy.logerr('Please upgrade schedule module. $ pip install schedule==0.6.0 --user')  # NOQA
         if 'stop' in app_schedule:
             stop_job = self._create_stop_job(name, app_name)  # NOQA
             try:
-                eval('schedule.{}.do(stop_job)'.format(app_schedule['stop']))
+                eval('schedule.{}.do(stop_job).tag(\'{}\')'.format(app_schedule['stop'],app['name']))
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
                eval('schedule.{}.do(stop_job).tag(\'{}\')'.format(app_schedule['stop'], app['name']))
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 4]
```diff
@@ -24,15 +31,61 @@ def __init__(self, robot_name, yaml_path, duration, update_duration):
...
         self.job_timer = rospy.Timer(rospy.Duration(duration), self._timer_cb)
         self.update_timer = rospy.Timer(
             rospy.Duration(update_duration), self._update_timer_cb)
+        self.pub_schedules = rospy.Publisher(
+            '~app_schedules', AppScheduleEntries, queue_size=1)
         self.sub = rospy.Subscriber(
             '/{}/application/app_status'.format(self.robot_name),
             AppStatus, self._sub_cb)
+        self.srv_add_entry = rospy.Service(
+            '~add_entry', AddEntry, self._handler_add_entry)
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
            '~add_entry', AddEntry, self._srv_add_entry_cb)
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 5]
```diff
@@ -24,15 +31,61 @@ def __init__(self, robot_name, yaml_path, duration, update_duration):
...
+        self.srv_remove_entry = rospy.Service(
+            '~remove_entry', RemoveEntry, self._handler_remove_entry)
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
            '~remove_entry', RemoveEntry, self._srv_remove_entry_cb)
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 6]
```diff
@@ -144,3 +201,16 @@ def _sub_cb(self, msg):
         else:
             # ERROR
             rospy.logerr('app_scheduler: {}'.format(msg.status))
+
+    def _handler_add_entry(self, req):
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
    def _srv_add_entry_cb(self, req):
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py [hunk 7]
```diff
@@ -144,3 +201,16 @@ def _sub_cb(self, msg):
...
+        self._add_entry(req.entry)
+        self._publish_app_schedules()
+        res = AddEntryResponse()
+        return res
+
+    def _handler_remove_entry(self, req):
```
Comment (knorth55) [inline app_scheduler/src/app_scheduler/app_scheduler_lib.py]:
```suggestion
    def _srv_remove_entry_cb(self, req):
```

**Review by knorth55 — CHANGES_REQUESTED** [review]

**Comment by knorth55** [comment]
Test failed in `flake8` linter.
You can check this with `flake8` + `hacking`.
```bash
pip install flake8 hacking
```

```
2021-12-03 08:06:59 [ERROR]   Found errors in [flake8] linter!
2021-12-03 08:06:59 [ERROR]   Error code: 1. Command output:
------
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:52:17: E126 continuation line over-indented for hanging indent
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:55:17: E123 closing bracket does not match indentation of opening bracket's line
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:70:22: E201 whitespace after '['
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:70:70: E202 whitespace before ']'
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:81:35: W601 .has_key() is deprecated, use 'in'
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:83:35: W601 .has_key() is deprecated, use 'in'
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:111:80: E501 line too long (99 > 79 characters)
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:111:86: E231 missing whitespace after ','
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:119:80: E501 line too long (101 > 79 characters)
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:119:88: E231 missing whitespace after ','
/github/workspace/app_scheduler/src/app_scheduler/app_scheduler_lib.py:216:1: W391 blank line at end of file
```

File: app_scheduler/src/app_scheduler/app_scheduler_lib.py (hunk 1 above)
Comment (sktometometo) [reply app_scheduler/src/app_scheduler/app_scheduler_lib.py]: Fixed.

**Review by knorth55 — APPROVED** [review]

**Comment by tkmtnt7000** [comment]
Today Fetch's morning kitchen demo did not start because jsk-app-scheduler didn't launch.
```
waiting for roscore to come up
waiting for app manager to come up
master has started, initiating launch
... logging to /home/fetch/.ros/log/0f2b205c-5790-11ec-b1ec-79b1ba52627e/roslaunch-fetch1075-19795.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
WARNING: disk usage in log directory [/home/fetch/.ros/log] is over 1GB.
It's recommended that you use the 'rosclean' command.

started roslaunch server http://fetch1075:44219/

SUMMARY
========

PARAMETERS
 * /app_scheduler/duration: 1
 * /app_scheduler/update_duration: 10
 * /app_scheduler/yaml_path: /var/lib/robot/ap...
 * /rosdistro: melodic
 * /rosversion: 1.14.11

NODES
  /
    app_scheduler (app_scheduler/app_scheduler)

ROS_MASTER_URI=http://localhost:11311

process[app_scheduler-1]: started with pid [20006]
Traceback (most recent call last):
  File "/home/fetch/ros/melodic/src/knorth55/app_manager_utils/app_scheduler/scripts/app_scheduler", line 5, in <module>
    from app_scheduler import AppScheduler
  File "/home/fetch/ros/melodic/devel/lib/python2.7/dist-packages/app_scheduler/__init__.py", line 34, in <module>
    exec(__fh.read())
  File "<string>", line 1, in <module>
  File "/home/fetch/ros/melodic/src/knorth55/app_manager_utils/app_scheduler/src/app_scheduler/app_scheduler_lib.py", line 14, in <module>
    from app_scheduler.msg import AppScheduleEntries
ImportError: No module named msg
[app_scheduler-1] process has died [pid 20006, exit code 1, cmd /home/fetch/ros/melodic/src/knorth55/app_manager_utils/app_scheduler/scripts/app_scheduler __name:=app_scheduler __log:=/home/
fetch/.ros/log/0f2b205c-5790-11ec-b1ec-79b1ba52627e/app_scheduler-1.log].
log file: /home/fetch/.ros/log/0f2b205c-5790-11ec-b1ec-79b1ba52627e/app_scheduler-1*.log
all processes on machine have died, roslaunch will exit
shutting down processing monitor...
... shutting down processing monitor complete
done
```
Probably `ImportError: No module named msg` occurs when the package name and the msg name is same. So I think the msg name `app_scheduler` has to be replaced by other name.
https://answers.ros.org/question/113671/catkin-package-cannot-find-own-message-type-python/

**Comment by knorth55** [comment]
We need to build the package again to generate new message files, I think

**Comment by tkmtnt7000** [comment]
I tried building the package again, but jsk-app-scheduler did not launch with the same error.
These new messages seem to be  generated as following.
```
fetch@fetch1075:~$ rosmsg list | grep app_scheduler
app_scheduler/AppSchedule
app_scheduler/AppScheduleEntries
app_scheduler/AppScheduleEntry
```

- What I tried
1. executing only `catkin build` and launching jsk-app-scheduler.
2. executing `catkin clean app_scheduler` and `catkin build app_scheduler` and launching jsk-app-scheduler.
3. reboot (This is not meaningful so much, I think)

**Comment by tkmtnt7000** [comment]
This problem is solved after building several times.
One of the possible causes is that the file referred to by import is different because the previous `app_scheduler.pyc` is still in effect. So when `app_schedule.pyc` is removed, jsk-app-scheduler successfully launches.
Thank you very much for advice.

**Pull request merged**